File: src/main/java/Emitter.java
import com.google.gson.Gson;

import java.io.PrintWriter;
import java.util.HashMap;
import java.util.Map;

public class Emitter {
    private PrintWriter writer;
    private Gson gson = new Gson();
    private int nextId = 1;

    public Emitter(PrintWriter writer) {
        this.writer = writer;
    }

    public int numElements() {
        return nextId;
    }

    public String emitVertex(String labelName, Map<String, Object> args) {
        return emit("vertex", labelName, args);
    }

    public String emitEdge(String labelName, Map<String, Object> args) {
        return emit("edge", labelName, args);
    }

    private String emit(String typeName, String labelName, Map<String, Object> args) {
        String id = String.format("%d", nextId++);

        writer.println(gson.toJson(Util.union(args, Util.mapOf(
                "id", id,
                "type", typeName,
                "label", labelName
        ))));

        return id;
    }
}

File: src/main/kotlin/lsifjava/Emitter.kt
package lsifjava

import java.io.PrintWriter

import com.google.gson.Gson

class Emitter(private val writer: PrintWriter) {
    private var nextId: Int = 1

    private val gson: Gson = Gson()

    fun numElements() = nextId

    fun emitVertex(labelName: String, args: Map<String, Any>) = emit("vertex", labelName, args)

    fun emitEdge(labelName: String, args: Map<String, Any>) = emit("edge", labelName, args)

    private fun emit(typeName: String, labelName: String, args: Map<String, Any>): String {
        val id = (nextId++).toString()

        val map = mutableMapOf<String, Any>(
            "id" to id,
            "type" to typeName,
            "label" to labelName
        )
        map.putAll(args)

        writer.println(gson.toJson(map))

        return id
    }
}